File: impls/rust/src/reader.rs
use std::{collections::HashMap, fmt::Display};

use crate::{
    tokenize::{Token, Tokenizer},
    value::Value,
};

#[derive(Debug)]
pub enum ParseError {
    EmptyInput,
    UnexpectedEof,
    UnterminatedString,
    EmptyKeyword,
    InvalidNumber,
    UnexpectedToken,
    InvalidStringEscape(char),
    InvalidMapKey(String),
}

impl Display for ParseError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // FIXME: better errors!!
        match self {
            ParseError::UnexpectedEof | ParseError::UnterminatedString => {
                write!(f, "unexpected end of input")
            }
            ParseError::InvalidNumber => write!(f, "invalid number"),
            ParseError::EmptyKeyword => write!(f, "empty keyword"),
            ParseError::UnexpectedToken => write!(f, "unexpected token"),
            ParseError::InvalidStringEscape(c) => write!(f, "invalid string escape: \\{}", c),
            ParseError::InvalidMapKey(k) => write!(f, "invalid map key '{}'", k),
            ParseError::EmptyInput => unreachable!(),
        }
    }
}

pub type ParseResult<T> = Result<T, ParseError>;

struct Reader<'a> {
    tokenizer: Tokenizer<'a>,
    next_token: Option<Token>,
}

impl<'a> Reader<'a> {
    fn new(input: &'a str) -> ParseResult<Self> {
        let mut tokenizer = Tokenizer::new(input);
        Ok(Self {
            next_token: Some(tokenizer.next_token().map_err(|e| match e {
                ParseError::UnexpectedEof => ParseError::EmptyInput,
                e => e,
            })?),
            tokenizer,
        })
    }
    fn next(&mut self) -> ParseResult<Token> {
        if let Some(token) = self.next_token.take() {
            Ok(token)
        } else {
            self.tokenizer.next_token()
        }
    }

    fn peek(&mut self) -> ParseResult<&Token> {
        if self.next_token.is_none() {
            self.next_token = Some(self.tokenizer.next_token()?);
        }
        Ok(self.next_token.as_ref().unwrap())
    }

    fn read_form(&mut self) -> ParseResult<Value> {
        let token = self.peek()?;
        match token {
            Token::LeftParen => Ok(Value::List(self.read_list(Token::RightParen)?)),
            Token::LeftBracket => Ok(Value::Vec(self.read_list(Token::RightBracket)?)),
            Token::LeftBrace => Ok(Value::Map(self.read_map()?)),
            Token::SingleQuote => {
                self.next().unwrap();
                self.read_reader_macro("quote")
            }
            Token::Backtick => {
                self.next().unwrap();
                self.read_reader_macro("quasiquote")
            }
            Token::Tilde => {
                self.next().unwrap();
                if matches!(self.peek(), Ok(&Token::At)) {
                    self.next().unwrap();
                    self.read_reader_macro("splice-unquote")
                } else {
                    self.read_reader_macro("unquote")
                }
            }
            Token::At => {
                self.next().unwrap();
                self.read_reader_macro("deref")
            }
            Token::Hat => {
                self.next().unwrap();
                let second = self.read_form()?;
                let first = self.read_form()?;
                Ok(Value::List(vec![
                    Value::Symbol("with-meta".into()),
                    first,
                    second,
                ]))
            }
            _ => self.read_atom(),
        }
    }

    fn read_reader_macro(&mut self, name: impl Into<String>) -> ParseResult<Value> {
        let content = self.read_form()?;
        Ok(Value::List(vec![Value::Symbol(name.into()), content]))
    }

    fn read_list(&mut self, terminator: Token) -> ParseResult<Vec<Value>> {
        self.next()?;
        let mut values = vec![];
        while self.peek()? != &terminator {
            values.push(self.read_form()?)
        }
        self.next()?;
        Ok(values)
    }

    fn read_map(&mut self) -> ParseResult<HashMap<String, Value>> {
        self.next()?;
        let mut map = HashMap::new();
        while self.peek()? != &Token::RightBrace {
            map.insert(
                self.read_form()?
                    .into_hash_map_key()
                    .map_err(|e| ParseError::InvalidMapKey(e.to_string()))?,
                self.read_form()?,
            );
        }
        self.next()?;
        Ok(map)
    }

    fn read_atom(&mut self) -> ParseResult<Value> {
        let value = self.next()?;
        Ok(match value {
            Token::Number(n) => Value::Number(n.parse().map_err(|_| ParseError::InvalidNumber)?),
            Token::Ident(value) if value == "true" => Value::Bool(true),
            Token::Ident(value) if value == "false" => Value::Bool(false),
            Token::Ident(value) if value == "nil" => Value::Nil,
            Token::Ident(value) => Value::Symbol(value),
            Token::Keyword(value) => Value::Keyword(value),
            Token::String(value) => Value::String(value),
            _ => return Err(ParseError::UnexpectedToken),
        })
    }
}

pub fn read_str(input: &str) -> ParseResult<Value> {
    Reader::new(input)?.read_form()
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn wat() {
        read_str("~@(1 2 3)").unwrap();
    }
}

File: impls/js2/reader.mjs
export class Vec extends Array {}

class Reader {
  constructor(tokens) {
    this.tokens = tokens;
    this.position = 0;
  }
  next() {
    if (!this.hasTokens()) {
      throw new Error("unexpected EOF");
    }
    return this.tokens[this.position++];
  }
  peek() {
    if (!this.hasTokens()) {
      throw new Error("unexpected EOF");
    }
    return this.tokens[this.position];
  }

  hasTokens() {
    return this.position < this.tokens.length;
  }

  read_form() {
    switch (this.peek()) {
      case "(":
        return this.read_list(")");
      case "[":
        return Vec.from(this.read_list("]"));
      case "{":
        return this.read_map();
      case "'":
        this.next();
        return [Symbol.for("quote"), this.read_form()];
      case "`":
        this.next();
        return [Symbol.for("quasiquote"), this.read_form()];
      case "~":
        this.next();
        return [Symbol.for("unquote"), this.read_form()];
      case "~@":
        this.next();
        return [Symbol.for("splice-unquote"), this.read_form()];
      case "^":
        this.next();
        var meta = this.read_form();
        return [Symbol.for("with-meta"), this.read_form(), meta];
      case "@":
        this.next();
        return [Symbol.for("deref"), this.read_form()];
      default:
        return this.read_atom();
    }
  }
  read_map() {
    const map = new Map();
    const list = this.read_list("}");
    for (let index = 0; index < list.length; index += 2) {
      map.set(list[index], list[index + 1]);
    }
    return map;
  }
  read_list(end) {
    this.next();
    let list = [];
    while (this.peek()[0] != end) {
      list.push(this.read_form());
    }
    this.next();
    return list;
  }
  read_atom() {
    const token = this.next();
    if (token.match(/^-?[0-9]+$/)) {
      return parseInt(token, 10); // integer
    } else if (token.match(/^-?[0-9][0-9.]*$/)) {
      return parseFloat(token, 10); // float
    } else if (token.match(/^"(?:\\.|[^\\"])*"$/)) {
      return token
        .slice(1, token.length - 1)
        .replace(/\\(.)/g, (_, c) => (c === "n" ? "\n" : c));
    } else if (token[0] === '"') {
      throw new Error("unexpected EOF");
    } else if (token[0] === ":") {
      return "\u029e" + token.slice(1);
    } else if (token === "nil") {
      return null;
    } else if (token === "true") {
      return true;
    } else if (token === "false") {
      return false;
    } else {
      return Symbol.for(token); // symbol
    }
  }
}

function tokenize(input) {
  const token_regex =
    /[\s,]*(~@|[\[\]{}()'`~^@]|"(?:\\.|[^\\"])*"?|;.*|[^\s\[\]{}('"`,;)]*)/g;
  let tokens = [];
  let match;
  while ((match = token_regex.exec(input)[1]) != "") {
    if (match[0] === ";") {
      continue;
    }
    tokens.push(match);
  }
  return tokens;
}

export function read_str(input) {
  const tokens = tokenize(input);
  const reader = new Reader(tokens);
  return reader.read_form();
}
